### PR TITLE
Homogeneization of train pipelines

### PR DESCRIPTION
**Pull request by 14thibea**
The goal of this PR is to homogeneize train pipelines for CNN as it has been done for autoencoders in previous pipeline.

### CODE REVIEW

**Comment by 14thibea** [comment]
The train functions are completely shared accross the different pipelines.
The test functions have been put together, maybe `clinicadl test` could be implemented to reproduce what we have done to evaluate performance on the test set for the paper ?

**Comment by 14thibea** [comment]
To be able to merge the pipelines functions specific to the initialization of the models were implemented.
Indeed for patch and roi, one init_state was saved before the execution of the folds and then the same init_state was used to initialize the training of the folds.
For image it was not case, each fold is initialized randomly independently.

File: clinicadl/clinicadl/tools/deep_learning/cnn_utils.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+
+
+def test(model, dataloader, use_cuda, criterion, mode="image"):
+    """
+    Computes the balanced accuracy of the model
+
+    :param model: the network (subclass of nn.Module)
+    :param dataloader: a DataLoader wrapping a dataset
+    :param use_cuda: if True a gpu is used
+    :param criterion: (loss) function to calculate the loss
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/deep_learning/cnn_utils.py]: Insert in the docstrings the possible choices for  the argument `mode`.
Change to Google-style.

File: clinicadl/clinicadl/tools/deep_learning/cnn_utils.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+from clinicadl.tools.deep_learning.iotools import check_and_clean
+from clinicadl.tools.deep_learning import EarlyStopping, save_checkpoint
+
+
+#####################
+# CNN train / test  #
+#####################
+
+def train(model, train_loader, valid_loader, criterion, optimizer, resume, log_dir, model_dir, options):
+    """
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/deep_learning/cnn_utils.py:19]: This function can probably be externally visible. I suggest to use the google-style for docstrings (to facilitate user adoption): short description, followed by _Args_, , _Returns_, _Raises_. (see [here an example](http://google.github.io/styleguide/pyguide.html#383-functions-and-methods))

File: clinicadl/clinicadl/tools/deep_learning/data.py [hunk 3]
```diff
@@ -592,12 +592,60 @@ def extract_roi_from_mri(image_tensor, left_is_odd):
     return extracted_roi
 
 
+def return_dataset(mode, input_dir, data_df, preprocessing, transformations, params, cnn_index=None):
+
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/deep_learning/data.py:611]: This is a nice feature (to regroup all dataloaders). I think a small docstrig is important here to explain inputs of this function. 
In the future, we can think of using the `clinica_file_reader` to make easy the creation of custom dataloaders, with other modalities of neuroimagerie data (referenced in issue #43)

File: clinicadl/clinicadl/train/train_multiCNN.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+from ..tools.deep_learning.models import transfer_learning, save_initialization, init_model
+from ..tools.deep_learning.data import (get_transforms,
+                                        load_data,
+                                        return_dataset)
+from ..tools.deep_learning.cnn_utils import train, soft_voting_to_tsvs
+from clinicadl.test.test_multiCNN import test_cnn
+
+
+__author__ = "Junhao Wen, Elina Thibeau-Sutre, Mauricio Diaz"
```
Comment (mdiazmel) [inline clinicadl/clinicadl/train/train_multiCNN.py]: These lines can be removed!

File: clinicadl/clinicadl/train/train_singleCNN.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+from ..tools.deep_learning.models import transfer_learning, save_initialization, init_model
+from ..tools.deep_learning.data import (get_transforms,
+                                        load_data,
+                                        return_dataset)
+from ..tools.deep_learning.cnn_utils import train
+from clinicadl.test.test_singleCNN import test_cnn
+
+
+__author__ = "Junhao Wen, Elina Thibeau-Sutre, Mauricio Diaz"
```
Comment (mdiazmel) [inline clinicadl/clinicadl/train/train_singleCNN.py]: To be removed.

File: clinicadl/clinicadl/train/train_singleCNN.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+__credits__ = ["Junhao Wen", "Elina Thibeau-Sutre", "Mauricio Diaz"]
+__license__ = "See LICENSE.txt file"
+__version__ = "0.1.0"
+__maintainer__ = "Junhao Wen, Elina Thibeau-Sutre, Mauricio Diaz"
+__email__ = "junhao.wen89@gmail.com, mauricio.diaz@inria.fr"
+__status__ = "Development"
+
+
+def train_single_cnn(params):
+
```
Comment (mdiazmel) [inline clinicadl/clinicadl/train/train_singleCNN.py:27]: Add docstring.

File: clinicadl/clinicadl/train/train_multiCNN.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+__credits__ = ["Junhao Wen", "Elina Thibeau-Sutre", "Mauricio Diaz"]
+__license__ = "See LICENSE.txt file"
+__version__ = "0.1.0"
+__maintainer__ = "Junhao Wen, Elina Thibeau-Sutre, Mauricio Diaz"
+__email__ = "junhao.wen89@gmail.com, mauricio.diaz@inria.fr"
+__status__ = "Development"
+
+
+def train_multi_cnn(params):
+
```
Comment (mdiazmel) [inline clinicadl/clinicadl/train/train_multiCNN.py:28]: Add docstring.

**Review by mdiazmel — COMMENTED** [review]
Hi @14thibea,
Thanks for your PR! There are big changes! Folder structure is more organized now that the  training functions are regrouped. 
Most of the comments below concern the docstrings.
Concerning a new task called `clinicadl test` I suppose that it will share a lot of things with `clinica classify`. There are two options: to provide `clinica test` or provide scripts using `clinica classify` that  execute the same functionality than `test`... Feel free to open a new issue to continue the discussion there.

File: clinicadl/clinicadl/train/train_singleCNN.py (hunk 5 above)
Comment (14thibea) [reply clinicadl/clinicadl/train/train_singleCNN.py]: Done in https://github.com/aramis-lab/AD-DL/pull/45/commits/bf452fc598385e907ba34d1174615029119eaa4a

File: clinicadl/clinicadl/train/train_multiCNN.py (hunk 4 above)
Comment (14thibea) [reply clinicadl/clinicadl/train/train_multiCNN.py]: Done in https://github.com/aramis-lab/AD-DL/pull/45/commits/bf452fc598385e907ba34d1174615029119eaa4a

File: clinicadl/clinicadl/tools/deep_learning/cnn_utils.py (hunk 1 above)
Comment (14thibea) [reply clinicadl/clinicadl/tools/deep_learning/cnn_utils.py]: Done in https://github.com/aramis-lab/AD-DL/pull/45/commits/9f4e3fd97f19a102c71c1967dcc58940c6c7ba3a

File: clinicadl/clinicadl/tools/deep_learning/cnn_utils.py (hunk 2 above)
Comment (14thibea) [reply clinicadl/clinicadl/tools/deep_learning/cnn_utils.py:19]: Done in https://github.com/aramis-lab/AD-DL/pull/45/commits/9f4e3fd97f19a102c71c1967dcc58940c6c7ba3a

File: clinicadl/clinicadl/train/train_multiCNN.py (hunk 7 above)
Comment (14thibea) [reply clinicadl/clinicadl/train/train_multiCNN.py:28]: Done in https://github.com/aramis-lab/AD-DL/pull/45/commits/86ce8739c926869c7b5420eccdcb11627a4e4582

File: clinicadl/clinicadl/train/train_singleCNN.py (hunk 6 above)
Comment (14thibea) [reply clinicadl/clinicadl/train/train_singleCNN.py:27]: Done in https://github.com/aramis-lab/AD-DL/pull/45/commits/86ce8739c926869c7b5420eccdcb11627a4e4582

File: clinicadl/clinicadl/tools/deep_learning/data.py (hunk 3 above)
Comment (14thibea) [reply clinicadl/clinicadl/tools/deep_learning/data.py:611]: Done in https://github.com/aramis-lab/AD-DL/pull/45/commits/86ce8739c926869c7b5420eccdcb11627a4e4582

**Comment by mdiazmel** [comment]
I ran all the training tests in local, everything is OK. @14thibea, ready to merge?

**Comment by 14thibea** [comment]
Yes !
